### PR TITLE
feat(list): add ability to pass dynamic attributes to list directive

### PR DESCRIPTION
**Pull request by rothm-caci**
List directive was modified to allow attributes to be passed from a parent scope to a directive with isolated scope compiled in the list directive using list-attr-* convention to map attributes, and with support for two-way binding.

template:
`<div list list-id="my-list" list-attr-my-item="myItem"></div>`

plugin:
`os.ui.list.add('my-list', '<my-directive item="myItem"></my-directive>');`

### CODE REVIEW

File: src/os/ui/list.js [hunk 1]
```diff
@@ -328,25 +350,33 @@ os.ui.ListCtrl.prototype.onChange = function(evt) {
...
-  var suffix = this.suffix || '';
+  const prefix = this.prefix || '';
+  const suffix = this.suffix || '';
 
   if (items) {
-    for (var i = 0, n = items.length; i < n; i++) {
-      var item = items[i];
+    let i = 0;
+    const n = items.length;
+    for (; i < n; i++) {
```
Comment (gregroyal) [inline src/os/ui/list.js]: why not define i on this line?

File: src/os/ui/list.js [hunk 2]
```diff
@@ -209,17 +210,32 @@ os.ui.Module.directive('list', [os.ui.listDirective]);
...
+      .reduce((acc, key) => {
+        let camel = key.replace(/^listAttr/, '');
+        camel = camel.charAt(0).toLowerCase() + camel.substr(1);
+        ctrl.attrs[camel] = scope.$eval(attr[key]);
+        scope.$watch(function() {
+          const kebab = key.replace(/([A-Z])/g, '-$1').toLowerCase();
+          return scope.$eval(element.attr(kebab));
+        }, function(newValue, oldValue) {
+          ctrl.attrs[camel] = newValue;
+          ctrl.update_();
```
Comment (gregroyal) [inline src/os/ui/list.js]: this looks like it re-compiles the directive anytime any of the parameters change

Comment (rothm-caci) [reply src/os/ui/list.js]: It now only compiles when !item.element, which is the first time it sees it.

File: src/os/ui/list.js (hunk 1 above)
Comment (rothm-caci) [reply src/os/ui/list.js]: Quite honestly, my IDE made the recommendation and I accepted it. I benchmarked it, and it seems to be marginally faster but significantly uglier.

File: src/os/ui/list.js [hunk 3]
```diff
@@ -328,25 +350,32 @@ os.ui.ListCtrl.prototype.onChange = function(evt) {
...
 
   if (items) {
-    for (var i = 0, n = items.length; i < n; i++) {
-      var item = items[i];
+    const n = items.length;
+    for (let i = 0; i < n; i++) {
+      const item = items[i];
+
+      if (item.scope && Object.keys(this.attrs).length) {
+        goog.object.extend(item.scope, this.attrs);
```
[the remark this answers is not in the document]
Comment (schmidtk) [reply src/os/ui/list.js]: We have been using `ol.obj.assign` which has its own polyfill if the native function doesn't exist, but [js-polyfills](https://github.com/inexorabletash/polyfill/blob/cefa69ee683474489685b126f3a74109735d8c76/es6.md#fundamental-objects) does provide it for us so we could switch to always using the native function.

[the remark this answers is not in the document]
Comment (rothm-caci) [reply src/os/ui/list.js]: Removed conformance requirement to in order to allow Object.assign in PR

**Pull request closed without merging**